File: benches/gate/tests/complete.rs
//! Does the compiled parser offer what the reference offers, at mise's scale?
//!
//! The rules a completion follows are usage-lib's, and a CLI's completions should not change
//! with the implementation that answers them. usage-lib reads a spec at run time and walks it;
//! usage-argv has only `&'static` tables — so the rules are reimplemented, and reimplemented
//! rules drift. Here both are asked the same question about mise's real spec and their answers
//! compared.
//!
//! Of eighteen lines tried across mise's tree, fifteen agree exactly — including sets of forty
//! and twenty-one candidates. The three that do not are here rather than quietly dropped,
//! because a parity test that only lists what passes is a test that says nothing:
//!
//! - `mise ⌶` — the reference runs the `run=` completion mise's root declares for `[TASK]`,
//!   which shells out to the CLI being completed and returns this repository's task list. A
//!   later stage answers those in-process; nothing here can produce them.
//! - `mise settings ⌶` — the reference offers all 273 settings from a `type="config_keys"`
//!   completion. Also a later stage: the reserved `type=` vocabulary is not read here yet.
//!
//! A fourth was a real gap, now closed: `flag "-p --path --file"` has a second long form that
//! `gen-shadow` dropped, so `mise use --⌶` was missing `--file`. The derive could express it
//! all along. The gap was invisible in the help comparison, which renders one long form per
//! flag, and showed up here because a completion offers every name a flag answers to.
//!
//! Each one is a difference between the *shadow* and the spec rather than between the two
//! completion implementations, which is the same standard the help comparison holds.

use usage::Spec as LibSpec;
use usage_argv::complete::{candidates, split, Shell};

fn mise_spec() -> LibSpec {
    include_str!("../../mise.usage.kdl")
        .parse()
        .expect("mise's spec should parse")
}

/// What each side offers for a line, as a sorted list of values.
///
/// Values only: the reference leaves a flag's description empty — its own source says
/// `TODO: get flag description` — while the compiled side has the help text in the metadata it
/// already carries. Offering it is a deliberate improvement rather than a drift, and comparing
/// values is what holds the part that must not differ.
fn both(line: &str) -> (Vec<String>, Vec<String>) {
    let cursor = line.len();
    let s = split(line, cursor, Shell::Bash);

    let ours = {
        let mut v: Vec<String> = candidates(shadow_mise::Cli::spec(), &s)
            .into_iter()
            .map(|c| c.value)
            .collect();
        v.sort();
        v.dedup();
        v
    };

    let theirs = {
        let spec = mise_spec();
        let mut v: Vec<String> = usage_cli::complete_candidates(&spec, &s.words, s.cword, "bash")
            .expect("the reference should answer")
            .into_iter()
            .map(|(value, _)| value)
            .collect();
        v.sort();
        v.dedup();
        v
    };

    (ours, theirs)
}

fn assert_same(line: &str) {
    let (ours, theirs) = both(line);
    assert_eq!(
        ours,
        theirs,
        "\n{line:?}\n  only ours: {:?}\n  only theirs: {:?}",
        ours.iter()
            .filter(|c| !theirs.contains(c))
            .collect::<Vec<_>>(),
        theirs
            .iter()
            .filter(|c| !ours.contains(c))
            .collect::<Vec<_>>(),
    );
    assert!(!ours.is_empty(), "{line:?} should offer something");
}

#[test]
fn the_subcommands_offered_are_the_reference_s() {
    // Including every alias each command answers to, and none of the hidden ones: the parse
    // table cannot tell those apart, since it must accept both, so offering the right set
    // depends on reading the metadata rather than the table.
    assert_same("mise plugins ");
    assert_same("mise plugins l");
    assert_same("mise config ");
}

#[test]
fn the_long_flags_offered_are_the_reference_s() {
    assert_same("mise --");
    // Every long form a flag answers to, including the second of `-p --path --file`.
    assert_same("mise use --");
    assert_same("mise install --");
    assert_same("mise run --");
    assert_same("mise config ls --");
    assert_same("mise tasks --");
    assert_same("mise plugins ls --");
    assert_same("mise upgrade --");
    assert_same("mise ls --");
}

#[test]
fn a_prefix_matching_no_declared_choice_is_answered_the_same_way() {
    // `mise activate zsx⌶` — the argument declares its whole set, so nothing matching means no
    // matches, not "here is the working directory". Both sides must agree, and both must be
    // empty: this is the case where a filtered-list test would have passed while the rule was
    // wrong, because the rule is about what the position *declares*.
    let spec = mise_spec();
    for line in ["mise activate zs", "mise activate zsx"] {
        let s = split(line, line.len(), Shell::Bash);
        let ours = usage_argv::complete::complete(shadow_mise::Cli::spec(), &s);
        let theirs = usage_cli::complete_candidates(&spec, &s.words, s.cword, "bash")
            .expect("the reference should answer");
        let theirs: Vec<String> = theirs.into_iter().map(|(v, _)| v).collect();
        let ours_values: Vec<String> = ours.candidates.iter().map(|c| c.value.clone()).collect();
        assert_eq!(ours_values, theirs, "{line:?}");
        assert_eq!(
            ours.files, None,
            "{line:?} declares its set, so no paths belong"
        );
    }
}

#[test]
fn the_short_flags_offered_are_the_reference_s() {
    // A lone dash offers both forms; a letter narrows to the flag that has it.
    assert_same("mise -");
    assert_same("mise use -");
    assert_same("mise use -g");
    assert_same("mise install -f");
}

/// Where the reference reads the directory, this says the shell should.
///
/// The one deliberate divergence in the completion path, so it is checked as an equivalence
/// rather than waived: for a line whose answer is paths, the reference's candidates must all be
/// entries that exist in the working directory, and ours must be the marker saying so. A
/// difference either way — a listing where we claim to know the answer, or a marker where the
/// reference had real candidates — fails.
#[test]
fn where_the_reference_lists_files_this_asks_the_shell_for_them() {
    use usage_argv::complete::{complete, Files};

    let spec = mise_spec();
    let cwd = std::env::current_dir().expect("a working directory");
    let entries: Vec<String> = std::fs::read_dir(&cwd)
        .expect("readable")
        .filter_map(|e| e.ok())
        .map(|e| e.file_name().to_string_lossy().to_string())
        .collect();
    assert!(
        !entries.is_empty(),
        "the fixture needs a non-empty directory"
    );

    // `mise trust ⌶` takes a `[CONFIG_FILE]`, which the reference completes as a path.
    for line in ["mise trust ", "mise config get --file "] {
        let s = split(line, line.len(), Shell::Bash);
        let ours = complete(shadow_mise::Cli::spec(), &s);
        let theirs = usage_cli::complete_candidates(&spec, &s.words, s.cword, "bash")
            .expect("the reference should answer");

        assert_eq!(
            ours.files,
            Some(Files::Any),
            "{line:?} should hand paths to the shell, got {ours:?}"
        );
        // Every name the reference offered is something in this directory — i.e. it answered
        // with a listing, which is the thing we are replacing rather than contradicting.
        for (value, _) in &theirs {
            let bare = value.trim_end_matches('/');
            assert!(
                entries.iter().any(|e| e == bare),
                "{line:?}: the reference offered {value:?}, which is not a directory entry — so \
                 it was not doing file completion and the marker is wrong here"
            );
        }
    }
}

File: go/argv/complete.go
package argv

import "strings"

// What could go where the cursor is.
//
// A completion is a parse of an unfinished command line, which is why it lives
// beside the parser rather than on top of it: the words before the cursor decide
// what may follow, and the rules that decide are the binding rules. Asking the
// parser rather than re-deriving them is what keeps what is *offered* and what is
// *accepted* from disagreeing — a completion advertising a flag the parser would
// refuse is worse than no completion at all.
//
// This is the position and the candidates. Turning candidates into the format a
// particular shell wants, and running the `complete` scripts a spec can declare,
// are separate jobs: the first is per-shell text, the second runs subprocesses,
// and neither belongs in a package whose whole claim is that it does not allocate.

// Position is what the cursor is standing in, after the words before it.
type Position struct {
	// Cmd is the command in scope: the deepest one the words selected.
	Cmd *Command
	// Chain is the commands the words passed through, root first, which is what
	// [ShortHelp] and the scope rules want.
	Chain []*Command
	// FlagsPossible is whether a dash-prefixed word here would still be read as a
	// flag. False past a `--`, and past the first value of an `automatic`
	// argument — there is no flag of *this* CLI to offer in either place.
	FlagsPossible bool
	// SubcommandsPossible is whether a word here could still name a subcommand.
	// False once a positional of this command has taken a word: the parser stops
	// descending there, so a later word matching a subcommand name is a value.
	SubcommandsPossible bool
	// AwaitingValue is a flag whose value the cursor is standing in, because the
	// last word was a flag that takes one and has not been given it. Nothing else
	// belongs here: the parser refuses a flag-like token in that place.
	AwaitingValue *Flag
	// Collecting is a variadic flag still claiming words. The next word would be
	// another of its values — so the positional after it is not offered — but a
	// flag-like token ends the collection and binds, so flags are.
	Collecting *Flag
	// NextArg is the positional a word here would fill, if any are left.
	NextArg *Arg
	// SeparatorSeen is whether a `--` has been typed. Narrower than
	// FlagsPossible, and what an argument requiring a separator is asking about.
	SeparatorSeen bool
	// HelpTopic is whether the word here names a command to *read about* rather
	// than one to run — after `help`, where nothing else belongs.
	HelpTopic bool
}

// Walk reads the words before the cursor and reports what the cursor is at.
//
// Errors are not failures here. A line being completed is by definition
// unfinished — a flag with no value yet, a word that names nothing yet — so a
// parse error means "the grammar runs out here", which is exactly the position
// being asked about. The walk stops at the first one and reports the state it
// reached, where a real parse must discard everything.
func Walk(root *Command, words []string) Position {
	p := New(root, words)
	chain := []*Command{root}
	var awaiting *Flag

	for p.Next() {
		if ev := p.Event(); ev.Kind == KindCommand {
			chain = append(chain, ev.Command)
		}
	}
	if err, ok := p.Err().(*Error); ok && err != nil {
		switch err.Code {
		// The one failure that says something about the cursor rather than about
		// the line: the last word was a flag that takes a value, so the cursor is
		// standing in it.
		case CodeMissingFlagValue:
			awaiting = err.Flag
		// `ex help config ⌶` asks which command to read about, and the answer is a
		// command under `config` — the one the help request already resolved. The
		// parser never descended into it, on purpose, so the position comes from
		// the request. Nothing else can be typed there: a topic takes no flags and
		// fills no argument.
		case CodeHelp:
			// SubcommandsPossible stays false: a topic is not descended into, and
			// the commands under it are offered by HelpTopic instead.
			return Position{Cmd: err.Cmd, Chain: chain, HelpTopic: true}
		}
	}

	return Position{
		Cmd:   p.Command(),
		Chain: chain,
		// A variadic flag still claiming words stands in the same place as a flag
		// waiting for its first value: the next word belongs to it, not to the
		// positional after it.
		AwaitingValue:       awaiting,
		Collecting:          p.Collecting(),
		FlagsPossible:       !p.FlagsStopped(),
		SubcommandsPossible: p.SubcommandsPossible(),
		NextArg:             p.PendingArg(),
		SeparatorSeen:       p.DoubleDashSeen(),
	}
}

// Kind of thing a candidate is, so a shell can decorate or filter them.
type CandidateKind uint8

const (
	// CandidateCommand is a subcommand name or alias.
	CandidateCommand CandidateKind = iota
	// CandidateFlag is a flag spelling.
	CandidateFlag
	// CandidateValue is one of a declared `choices` list.
	CandidateValue
)

// Candidate is one thing that could be typed where the cursor is.
type Candidate struct {
	Kind CandidateKind
	// Value is the text to insert.
	Value string
	// Describe is the one-line help, where there is any. A shell that can show a
	// description beside a completion uses it; one that cannot ignores it.
	Describe string
}

// Candidates is everything that could go at a position, given a partial word.
//
// `partial` is what the user has typed of the current word, and filtering happens
// here rather than in the shell so that every shell agrees about what matches.
func Candidates(pos Position, partial string, help HelpTable, meta Metadata) []Candidate {
	var out []Candidate
	add := func(kind CandidateKind, value, describe string) {
		if strings.HasPrefix(value, partial) {
			out = append(out, Candidate{Kind: kind, Value: value, Describe: describe})
		}
	}

	// A value the cursor is standing in takes the position entirely: nothing else
	// belongs where a flag is waiting for its argument, because the parser refuses
	// a flag-like token there.
	if pos.AwaitingValue != nil {
		for _, c := range choicesFor(pos.AwaitingValue.Key, meta) {
			add(CandidateValue, c, "")
		}
		return out
	}

	// Commands, and their aliases: the parser accepts either, and a completion
	// that hides an alias makes it undiscoverable. A hidden command binds and is
	// not advertised — the same rule the help pages follow, and the reason `hide`
	// exists at all.
	//
	// The same list after `help`, because `findNamed` resolves a topic by name or
	// alias exactly as it resolves a command to run.
	commands := func() {
		for _, sub := range subcommandsOf(pos.Cmd) {
			h := help.Lookup(sub.Key)
			if h != nil && h.Hide {
				continue
			}
			add(CandidateCommand, sub.Name, describe(sub.Key, help))
			if h != nil {
				for _, alias := range h.VisibleAliases {
					add(CandidateCommand, alias, describe(sub.Key, help))
				}
			}
		}
	}

	// A help topic is a question, not an invocation: only command names belong.
	if pos.HelpTopic {
		commands()
		return out
	}

	// A variadic flag that has already taken a value is a weaker claim than a flag
	// waiting for its first: another word goes to the variadic, but a flag-like
	// one ends the collection and binds. So its values are offered here *and* the
	// flags below are — what is not offered is anything a plain word could not be:
	// a subcommand, or the positional the variadic is standing in front of.
	collecting := pos.Collecting != nil
	if collecting {
		for _, c := range choicesFor(pos.Collecting.Key, meta) {
			add(CandidateValue, c, "")
		}
	}

	// Only while descent is still possible. Once a positional has taken a word the
	// parser stops matching subcommands, and a name offered there would be bound
	// as a value or refused outright.
	if pos.SubcommandsPossible && !collecting {
		commands()
	}

	// Flags, only where one could still be typed *and* the user has started typing
	// one. The reference offers no flags for a bare cursor — `ex ⌶` lists
	// subcommands, `ex -⌶` lists both forms, `ex --⌶` the longs — and checked
	// against `usage complete-word` rather than assumed. Offering them anyway made
	// every position look answered, which is also what decides whether the shell
	// should fall back to paths.
	//
	// The prefix filter below would narrow them to the same set; what this changes
	// is the empty prefix, where it would not.
	if pos.FlagsPossible && strings.HasPrefix(partial, "-") {
		for _, s := range flagsInScope(pos.Chain) {
			if h := help.Lookup(s.flag.Key); h != nil && h.Hide {
				continue
			}
			// Negations included: flagsInScope works out which spellings are still
			// this flag's, and a negation is one of them.
			for _, form := range s.forms {
				add(CandidateFlag, form, describe(s.flag.Key, help))
			}
		}
	}

	// And the values a positional will only accept — unless it is one that reads
	// only after a `--` and no separator has been typed. Offering them there
	// produces a command line the parser answers with
	// `arg_requires_double_dash`, which is the exact failure this design exists
	// to prevent.
	if pos.NextArg != nil && !collecting &&
		!(pos.NextArg.DoubleDash == DoubleDashRequired && !pos.SeparatorSeen) {
		for _, c := range choicesFor(pos.NextArg.Key, meta) {
			add(CandidateValue, c, "")
		}
	}
	return out
}

// inScope is a flag a page or a completion may offer, and the spellings still
// left to it.
type inScope struct {
	flag  *Flag
	forms []string
}

// flagsInScope is this command's own flags, then any ancestor's globals, each
// with the spellings nothing nearer has taken.
//
// Per spelling, not per flag. A flag answers to several forms, and a nearer
// command reclaiming `--jobs` leaves an inherited `-j` and `--workers` binding —
// dropping the whole inherited flag would hide spellings the parser still
// accepts. That is the same rule the help pages follow, for the same reason.
func flagsInScope(chain []*Command) []inScope {
	if len(chain) == 0 {
		return nil
	}
	everyForm := everyFormInScope(chain)
	var taken, takenNegations []string
	var out []inScope

	// Nearest first, which is the order the parser resolves in, so that "nothing
	// nearer has taken it" is just "not seen yet".
	offer := func(f *Flag) {
		var left []string
		for _, form := range formsOf(f) {
			if !has(taken, form) {
				left = append(left, form)
			}
		}
		// A negation is a spelling like any other, and it loses to a long anywhere
		// in scope rather than only to a nearer one — see negationSurvives, which
		// the pages use for the same decision.
		//
		// Not twice, though. A flag may spell its negation the same as its own long
		// — `flag "--no-color" negate="--no-color"` — and then it is already in the
		// list. usage-lib prints that flag as `--no-color / --no-color`, so the page
		// says it twice on purpose and matching the reference means keeping that;
		// a completion is a list of things to type, and the same thing twice is
		// just a repeated row.
		// `!has(taken, n)` before the exemption inside negationSurvives: that
		// exemption is for a flag whose *own* long is spelled like its negation,
		// and it must not reach past something nearer that has claimed the word.
		// A child declaring `--x` takes it from an inherited global that answers to
		// `--x` both ways.
		if n := negationOf(f); n != "" && !has(left, n) && !has(taken, n) &&
			negationSurvives(f, n, takenNegations, everyForm) {
			left = append(left, n)
		}
		// Claimed whether or not anything is left: a spelling this flag answers to
		// is not available to something farther away either.
		taken = append(taken, formsOf(f)...)
		if n := negationOf(f); n != "" {
			takenNegations = append(takenNegations, n)
		}
		if len(left) > 0 {
			out = append(out, inScope{flag: f, forms: left})
		}
	}

	for _, f := range chain[len(chain)-1].Flags {
		offer(f)
	}
	for i := len(chain) - 2; i >= 0; i-- {
		for _, f := range chain[i].Flags {
			if f.Global {
				offer(f)
			}
		}
	}
	return out
}

func subcommandsOf(cmd *Command) []*Command {
	if cmd == nil {
		return nil
	}
	return cmd.Subcommands
}

func choicesFor(key uint64, meta Metadata) []string {
	if m := meta.Lookup(key); m != nil {
		return m.Choices
	}
	return nil
}

func describe(key uint64, help HelpTable) string {
	if h := help.Lookup(key); h != nil {
		// Whole, breaks included. Putting a description on one line is the
		// renderer's job — see oneLine — because it is the line-based protocols
		// that need it, and collapsing there keeps both halves of a two-line
		// description instead of dropping the second.
		return h.Short
	}
	return ""
}
